### PR TITLE
Move `DataCollatorForMultipleChoice` from the docs to the package

### LINKED ISSUE #34671: `DataCollatorForMultipleChoice` exists in the docs but not in the package
### Feature request

Move the `DataCollatorForMultipleChoice` implementation from the docs to `transformers.data.data_collator.py`.

### Motivation

The [`transformers` docs](https://huggingface.co/docs/transformers/tasks/multiple_choice) provide all collator code needed to run `ForMultipleChoice` fine-tuning for datasets like SWAG. The docs say that *"`transformers` doesn’t have a data collator for multiple choice, so you’ll need to adapt the `DataCollatorWithPadding` to create a batch of examples"*, but... why? Why not add the code given in the docs to the package? It's the same repo...

https://github.com/huggingface/transformers/blob/a06a0d12636756352494b99b5b264ac9955bc735/docs/source/en/tasks/multiple_choice.md?plain=1#L115-L160

### Your contribution

None, the code is already there.

### PR DESCRIPTION
**Pull request by bauwenst**
# What does this PR do?
Fixes #34671 by moving the implementations for `DataCollatorForMultipleChoice` that appear in the docs (copied across four different languages) into the `transformers` package.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request) "Pull Request" section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@LysandreJik asked me to contribute this. @ArthurZucker may also be interested, and @stevhliu because the implementations came from the docs.

### CODE REVIEW

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -532,12 +532,89 @@ def _numpy_collate_batch(examples, tokenizer, pad_to_multiple_of: Optional[int]
...
-def tolist(x):
-    if isinstance(x, list):
-        return x
-    elif hasattr(x, "numpy"):  # Checks for TF tensors without needing the import
-        x = x.numpy()
-    return x.tolist()
+@dataclass
+class DataCollatorForMultipleChoice(DataCollatorMixin):
+    """
+    Data collator that will dynamically pad the inputs for multiple choice received.
```
Comment (stevhliu) [inline src/transformers/data/data_collator.py:538]:
```suggestion
    Data collator that dynamically pads the inputs for multiple choice received.
```

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -532,12 +532,89 @@ def _numpy_collate_batch(examples, tokenizer, pad_to_multiple_of: Optional[int]
...
+
+    Args:
+        tokenizer ([`PreTrainedTokenizer`] or [`PreTrainedTokenizerFast`]):
+            The tokenizer used for encoding the data.
+        padding (`bool`, `str` or [`~utils.PaddingStrategy`], *optional*, defaults to `True`):
+            Select a strategy to pad the returned sequences (according to the model's padding side and padding index)
+            among:
```
Comment (stevhliu) [inline src/transformers/data/data_collator.py:545]:
```suggestion
            Select a strategy to pad the returned sequences according to the model's padding side and padding index:
```

File: src/transformers/data/data_collator.py [hunk 3]
```diff
@@ -532,12 +532,89 @@ def _numpy_collate_batch(examples, tokenizer, pad_to_multiple_of: Optional[int]
...
+
+            - `True` or `'longest'`: Pad to the longest sequence in the batch (or no padding if only a single sequence
+              if provided).
```
Comment (stevhliu) [inline src/transformers/data/data_collator.py:548]:
```suggestion
              is provided).
```

File: src/transformers/data/data_collator.py [hunk 4]
```diff
@@ -532,12 +532,89 @@ def _numpy_collate_batch(examples, tokenizer, pad_to_multiple_of: Optional[int]
...
+            - `'max_length'`: Pad to a maximum length specified with the argument `max_length` or to the maximum
+              acceptable input length for the model if that argument is not provided.
+            - `False` or `'do_not_pad'` (default): No padding (i.e., can output a batch with sequences of different
+              lengths).
+        max_length (`int`, *optional*):
+            Maximum length of the returned list and optionally padding length (see above).
+        pad_to_multiple_of (`int`, *optional*):
+            If set will pad the sequence to a multiple of the provided value.
```
Comment (stevhliu) [inline src/transformers/data/data_collator.py:556]:
```suggestion
            Pad the sequence to a multiple of the provided value.
```

File: docs/source/en/tasks/multiple_choice.md [hunk 5]
```diff
@@ -112,96 +112,7 @@ To apply the preprocessing function over the entire dataset, use 🤗 Datasets [
...
-...             return_tensors="tf",
-...         )
-
-...         batch = {k: tf.reshape(v, (batch_size, num_choices, -1)) for k, v in batch.items()}
-...         batch["labels"] = tf.convert_to_tensor(labels, dtype=tf.int64)
-...         return batch
-```
-</tf>
-</frameworkcontent>
+To create a batch of examples, it's more efficient to *dynamically pad* the sentences to the longest length in a batch during collation, instead of padding the whole dataset to the maximum length. [`DataCollatorForMultipleChoice`] flattens all the model inputs, applies padding, and then unflattens the results.
```
Comment (stevhliu) [inline docs/source/en/tasks/multiple_choice.md:115]: May be useful to add a code snippet that instantiates the new `DataCollatorForMultipleChoice` class :)

**Review by stevhliu — COMMENTED** [review]
Really nice, this is a lot cleaner! You can also run `make style` to try and fix some of the failing CI tests.

File: docs/source/en/tasks/multiple_choice.md (hunk 5 above)
Comment (bauwenst) [reply docs/source/en/tasks/multiple_choice.md:115]: Yeah, I was thinking that perhaps a `from transformers import DataCollatorForMultipleChoice` was missing somewhere (it would be if you concatenated all the snippets of the doc into a hypothetical training script), but it seemed kind of awkward to have just an import statement and then moving on to the next section immediately.

**Review by LysandreJik — APPROVED** [review]
Great, thank you! 

@stevhliu, feel free to merge once you're happy with the docs

**Comment by LysandreJik** [comment]
Could you just run the code quality tool to ensure that the code quality passes? You can install them with the following, from the root of your clone:
```
pip install -e ".[quality]"
```
And then run them with:
```
make fixup
```

File: docs/source/en/tasks/multiple_choice.md (hunk 5 above)
Comment (bauwenst) [reply docs/source/en/tasks/multiple_choice.md:115]: The `Trainer` constructor in the docs is already written in the form `Trainer(..., collator=DataCollatorForMultipleChoice(tokenizer=tokenizer), ...)` hence why I didn't include a separate variable assignment.

**Comment by bauwenst** [comment]
> Could you just run the code quality tool to ensure that the code quality passes? You can install them with the following, from the root of your clone:

@LysandreJik I ran both `make style` and `make fixup` and have committed all the automatic changes. There is nothing else I can do on this PR, as far as I can see.

**Review by ArthurZucker — COMMENTED** [review]
sorry for the delay Do you need help getting the CI green? You need to rebase as well!

**Comment by bauwenst** [comment]
@ArthurZucker Yes, as I wrote before, I thought I had run everything I needed to (`make style` and `make fixup`). I don't know what else to do in order to complete this.

**Comment by ArthurZucker** [comment]
Thanks @bauwenst 🤗
